documentation/blog/2025-12-15-code-mode-mcp/index.md

```diff
@@ -0,0 +1,62 @@
+---
+title: "Code Mode MCP in goose"
+description: An emerging approach to MCP tool calling gets an open source implementation in goose
+authors:
+    - alexhancock
+---
+
+![code mode MCP in goose!](header-image.jpg)
+
+# Code Mode MCP
+
+There is an emerging approach to MCP tool calling referred to as "sandbox mode" or "code mode". These ideas were initially
+presented by Cloudflare in their [Code Mode: the better way to use MCP](https://blog.cloudflare.com/code-mode/) post and Anthropic
+in their [Code execution with MCP: Building more efficient agents](https://www.anthropic.com/engineering/code-execution-with-mcp)
+posts. Since the approach and the benefits are clearly laid out in those posts I will summarize them here.
+
+## The approach
+
+### Summary
+
+* Instead of exposing the tools directly to the model an MCP client application can:
+    * Generate a programmatic interface to these same tools (typically JS or TS powered)
+    * Provide a limited set of tools to the model (search available modules/tool source code, read the source code for a tool, and then a tool to execute some code)
+    * Run the code the model generates to call the programmatic API in a sandboxed environment for safety
+
+### Benefits
+
+* The model can progressively discover relevant tools, without all server and tool definitions in the context window from the beginning
+* The model can chain tool call results into inputs to further tool calls without the intermediate results needing to flow back to the model, costing tokens and exposing data which is potentially sensitive to the model unecessarily
+* The models pre-training datasets have made them very efficient at analyzing large programmatic APIs and deciding how to usefully call them, as compared to having been trained only on contrived examples of MCP tool calling
+
+## In goose
+
+In v1.17.0 of goose, we've introduced an open source implementation of this idea in a new platform extension called: Code Execution.
```

Comment:
this is not out yet..so we should make it clear to people that it's coming up. 

Because the blog is going to merge and go live immediately..and people will say where is code mode, but we're all on 1.16.1